server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -56,9 +56,23 @@ public class MapperService extends AbstractIndexComponent implements Closeable {
      */
     public enum MergeReason {
         /**
-         * Pre-flight check before sending a mapping update to the master
+         * Pre-flight check before sending a dynamic mapping update to the master
          */
-        MAPPING_UPDATE_PREFLIGHT,
+        MAPPING_AUTO_UPDATE_PREFLIGHT {
+            @Override
+            public boolean isAutoUpdate() {
+                return true;
+            }
+        },
+        /**
+         * Dynamic mapping updates
+         */
+        MAPPING_AUTO_UPDATE {
```

Comment:
Should this not also be used by `ParsedDocument.addDynamicMappingsUpdate`?

Reply:
Which merge reason to use in `ParsedDocument.addDynamicMappingsUpdate` doesn't matter much. Maybe we should even just ignore the incoming mapping in `addDynamicMappingsUpdate` if it's set already.

The only case were the method is called production code is to ensure that there's an empty mapping during `prepareIndex`:

https://github.com/elastic/elasticsearch/blob/79b7dbb5441f0370e893bd3cd86b4031884abbbb/server/src/main/java/org/elasticsearch/index/shard/IndexShard.java#L1008-L1019

So the merge reason doesn't have any effect as the incoming mapping is always empty. Also, the preflight check and then the mapping update in the master node happens after `addDynamicMappingsUpdate` which do take the correct merge reason into account and potentially truncate the dynamic mappings update.

Reply:
I agree to all of that, but still without fixing that, I think this PR should change the reason to be auto mapping update for consistency?

Reply:
Makes sense. I've updated the PR.